File: src/server/run_manager.py
"""Run manager for ThreatForest pipeline executions.

Bridges the web UI to the ThreatForest orchestrator by managing run lifecycle,
spawning background threads for pipeline execution, and routing progress events
to asyncio queues that WebSocket handlers consume.
"""

from __future__ import annotations

import asyncio
import json
import logging
import threading
import time
import uuid
from datetime import datetime, timezone
from pathlib import Path
from typing import Any, Callable, Protocol

from server.models import RunConfig, RunState
from server.scan_control import ScanControl

logger = logging.getLogger(__name__)


# ---------------------------------------------------------------------------
# Interaction request for interviewer HITL
# ---------------------------------------------------------------------------

class InteractionRequest:
    """Pending interviewer question waiting for a user response."""

    def __init__(self, run_id: str, reason: dict[str, Any]) -> None:
        self.run_id = run_id
        self.reason = reason
        self.response: str | None = None
        self._event = threading.Event()

    def wait(self) -> None:
        self._event.wait()

    def respond(self, text: str | None) -> None:
        self.response = text
        self._event.set()


# ---------------------------------------------------------------------------
# Progress event helpers
# ---------------------------------------------------------------------------

class ProgressEvent:
    """Lightweight progress event pushed through the queue to WebSocket clients."""

    def __init__(
        self,
        *,
        event_type: str,
        stage: str,
        percentage: float,
        message: str,
        details: dict[str, Any] | None = None,
    ) -> None:
        self.event_type = event_type
        self.stage = stage
        self.percentage = percentage
        self.message = message
        self.details = details or {}

    def to_dict(self) -> dict[str, Any]:
        return {
            "type": self.event_type,
            "stage": self.stage,
            "percentage": self.percentage,
            "message": self.message,
            "details": self.details,
            "server_ts": time.time() * 1000,  # epoch ms for client-side timing
        }


# ---------------------------------------------------------------------------
# Orchestrator protocol — pluggable so the real engine is optional
# ---------------------------------------------------------------------------

class OrchestratorExecutor(Protocol):
    """Protocol for executing a ThreatForest pipeline run.

    Implementations receive the run configuration, a progress callback, and an
    optional ``ScanControl``.  They should return a dict on normal completion or
    when interrupted.  The dict always contains ``"status"`` (``"complete"``,
    ``"pause"``, or ``"stop"``); a normal completion also includes
    ``"output_dir"`` and ``"app_id"``.
    """

    def __call__(
        self,
        config: RunConfig,
        progress_callback: Callable[[ProgressEvent], None],
        scan_control: Any | None = None,
        interaction_fn: Any | None = None,
    ) -> dict[str, str]:
        ...


# ---------------------------------------------------------------------------
# RunManager
# ---------------------------------------------------------------------------

class RunManager:
    """Manages ThreatForest pipeline runs and their progress event queues.

    Each run gets a unique ID, an ``asyncio.Queue`` for streaming progress
    events to WebSocket clients, a ``ScanControl`` for pause/stop signalling,
    and a background thread that drives the orchestrator.
    """

    def __init__(self, executor: OrchestratorExecutor | None = None) -> None:
        self.active_runs: dict[str, RunState] = {}
        self._queues: dict[str, asyncio.Queue[dict[str, Any]]] = {}
        self._loops: dict[str, asyncio.AbstractEventLoop] = {}
        self._controls: dict[str, ScanControl] = {}
        self._pending_interactions: dict[str, InteractionRequest] = {}
        self._event_history: dict[str, list[dict[str, Any]]] = {}
        self._executor = executor

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    def start_run(self, config: RunConfig) -> str:
        """Validate config, create a new run, and spawn the orchestrator.

        Returns the generated ``run_id``.

        Raises
        ------
        FileNotFoundError
            If ``config.project_path`` does not exist.
        NotADirectoryError
            If ``config.project_path`` exists but is not a directory.
        RuntimeError
            If no orchestrator executor has been configured.
        """
        project = Path(config.project_path)
        if not project.is_absolute():
            # Resolve relative paths against the repo root (src/../..)
            _repo_root = Path(__file__).resolve().parent.parent.parent
            project = _repo_root / project
        if not project.exists():
            raise FileNotFoundError(
                f"Project path does not exist: {config.project_path}"
            )
        if not project.is_dir():
            raise NotADirectoryError(
                f"Project path is not a directory: {config.project_path}"
            )

        if self._executor is None:
            raise RuntimeError(
                "No orchestrator executor configured. "
                "Provide one via RunManager(executor=...)."
            )

        run_id = uuid.uuid4().hex
        now = datetime.now(tz=timezone.utc).isoformat()

        state = RunState(
            run_id=run_id,
            status="pending",
            config=config,
            started_at=now,
        )
        self.active_runs[run_id] = state

        # Create the asyncio queue and capture the current event loop
        queue: asyncio.Queue[dict[str, Any]] = asyncio.Queue()
        self._queues[run_id] = queue

        try:
            loop = asyncio.get_running_loop()
        except RuntimeError:
            loop = None
        self._loops[run_id] = loop  # type: ignore[assignment]

        # Create a ScanControl for this run
        control = ScanControl()
        self._controls[run_id] = control

        # Spawn background thread
        thread = threading.Thread(
            target=self._execute_in_thread,
            args=(run_id, config),
            daemon=True,
        )
        thread.start()

        return run_id

    def get_progress_queue(self, run_id: str) -> asyncio.Queue[dict[str, Any]]:
        """Return the event queue for *run_id*.

        Raises
        ------
        KeyError
            If *run_id* is not known.
        """
        try:
            return self._queues[run_id]
        except KeyError:
            raise KeyError(f"Unknown run_id: {run_id}") from None

    def pause_run(self, run_id: str) -> None:
        """Request the executor to pause after the current stage completes.

        Raises
        ------
        KeyError
            If *run_id* is not known.
        RuntimeError
            If the run is not in a pausable state.
        """
        state = self.active_runs.get(run_id)
        if state is None:
            raise KeyError(f"Unknown run_id: {run_id}")
        if state.status not in ("pending", "running"):
            raise RuntimeError(
                f"Run {run_id} cannot be paused (current status: {state.status})"
            )
        control = self._controls.get(run_id)
        if control is None:
            raise RuntimeError(f"No control object for run {run_id}")
        control.request_pause()

    def stop_run(self, run_id: str) -> None:
        """Request the executor to stop after the current stage completes.

        If the run is already paused (executor has exited) this simply updates
        the RunState status to ``"stopped"`` without needing to signal anything.

        Raises
        ------
        KeyError
            If *run_id* is not known.
        RuntimeError
            If the run is in a terminal state and cannot be stopped.
        """
        state = self.active_runs.get(run_id)
        if state is None:
            raise KeyError(f"Unknown run_id: {run_id}")

        if state.status == "paused":
            # Executor thread has already exited; just flip the status.
            state.status = "stopped"
            state.completed_at = datetime.now(tz=timezone.utc).isoformat()
            # Remove pause_state.json so this run no longer appears as resumable
            control = self._controls.get(run_id)
            if control and control.run_dir:
                pause_file = Path(control.run_dir) / "pause_state.json"
                if pause_file.is_file():
                    pause_file.unlink()
            return

        if state.status not in ("pending", "running"):
            raise RuntimeError(
                f"Run {run_id} cannot be stopped (current status: {state.status})"
            )
        control = self._controls.get(run_id)
        if control is None:
            raise RuntimeError(f"No control object for run {run_id}")
        control.request_stop()

    def resume_run(self, run_id: str) -> str:
        """Create a new run that resumes from the last completed stage.

        Reads ``pause_state.json`` written by the executor when it was paused
        or stopped, rebuilds a ``RunConfig`` that skips already-completed nodes,
        and starts a fresh run that reuses the same run directory (so state
        files are available).

        Returns the new ``run_id``.

        Raises
        ------
        KeyError
            If *run_id* is not known.
        RuntimeError
            If the run has no pause state or is in a non-resumable status.
        """
        state = self.active_runs.get(run_id)
        if state is None:
            raise KeyError(f"Unknown run_id: {run_id}")

        if state.status not in ("paused", "stopped"):
            raise RuntimeError(
                f"Run {run_id} cannot be resumed (current status: {state.status})"
            )

        # Locate the run directory via ScanControl (set by executor on startup)
        control = self._controls.get(run_id)
        run_dir_str = control.run_dir if control else None
        if not run_dir_str:
            raise RuntimeError(
                f"Run directory not found for run {run_id}. "
                "The server may have restarted — please start a new run."
            )

        pause_file = Path(run_dir_str) / "pause_state.json"
        if not pause_file.is_file():
            raise RuntimeError(
                f"No pause_state.json found in {run_dir_str}. "
                "Cannot resume this run."
            )

        pause_data = json.loads(pause_file.read_text(encoding="utf-8"))
        config_data = pause_data.get("config", {})
        completed_nodes: list[str] = pause_data.get("completed_nodes", [])

        new_config = RunConfig(
            project_path=config_data.get("project_path", state.config.project_path),
            threat_source=config_data.get("threat_source", state.config.threat_source),
            threat_file_path=config_data.get("threat_file_path", state.config.threat_file_path),
            # Tell the executor to reuse the existing run directory
            resume_run_dir=run_dir_str,
            # Tell the graph to skip nodes whose outputs are already on disk
            skip_nodes=completed_nodes,
        )

        return self.start_run(new_config)

    def set_pending_interaction(self, run_id: str, req: InteractionRequest) -> None:
        self._pending_interactions[run_id] = req

    def submit_interaction_response(self, run_id: str, text: str | None) -> None:
        req = self._pending_interactions.get(run_id)
        if req is None:
            raise KeyError(f"No pending interaction for run {run_id}")
        req.respond(text)

    def clear_pending_interaction(self, run_id: str) -> None:
        self._pending_interactions.pop(run_id, None)

    def get_event_history(self, run_id: str) -> list[dict[str, Any]]:
        """Return a copy of all progress events recorded for *run_id*."""
        return list(self._event_history.get(run_id, []))

    def cleanup_run(self, run_id: str) -> None:
        """Remove the event queue and loop reference for a finished run.

        Called by the WebSocket handler after the connection closes to
        prevent memory leaks.  The ``active_runs`` entry is kept so that
        the dashboard endpoint can still serve results.  The ScanControl
        is kept so that ``resume_run`` can locate the run directory.

        For active runs the queue and event history are preserved so that
        a reconnecting client can replay the full history.
        """
        state = self.active_runs.get(run_id)
        is_terminal = state is not None and state.status in ("complete", "failed", "stopped")
        if is_terminal:
            self._queues.pop(run_id, None)
            self._loops.pop(run_id, None)
            self._event_history.pop(run_id, None)
        # For active runs: keep queue + history intact for reconnection

    # ------------------------------------------------------------------
    # Background execution
    # ------------------------------------------------------------------

    def _execute_in_thread(self, run_id: str, config: RunConfig) -> None:
        """Run the orchestrator in a background thread.

        Progress events are pushed to the asyncio queue so that the
        WebSocket handler can forward them to the client.
        """
        state = self.active_runs[run_id]
        state.status = "running"
        queue = self._queues[run_id]
        loop = self._loops.get(run_id)
        control = self._controls.get(run_id)

        def _push_event(event: ProgressEvent) -> None:
            """Thread-safe helper to enqueue a progress event."""
            payload = event.to_dict()
            self._event_history.setdefault(run_id, []).append(payload)
            if loop is not None and loop.is_running():
                loop.call_soon_threadsafe(queue.put_nowait, payload)
            else:
                queue.put_nowait(payload)

        def _web_interaction_fn(interrupts):
            """Handle interviewer/scanner-review interrupts by routing to the web UI."""
            for interrupt in interrupts:
                reason = interrupt.reason or {}
                req = InteractionRequest(run_id, reason)
                self.set_pending_interaction(run_id, req)

                # Route to the correct UI stage based on phase
                phase = reason.get("phase", "interviewer")
                if phase == "scanner_review":
                    stage_label = "Repository Analysis"
                elif phase == "threat_review":
                    stage_label = "Threat Review"
                else:
                    stage_label = "Context Validation"

                _push_event(ProgressEvent(
                    event_type="awaiting_input",
                    stage=stage_label,
                    percentage=50,
                    message=reason.get("message", "The interviewer has questions for you."),
                    details=reason,
                ))

                req.wait()
                self.clear_pending_interaction(run_id)

                if req.response is None:
                    return None

                return [{"interruptResponse": {"interruptId": interrupt.id, "response": req.response}}]
            return None

        try:
            assert self._executor is not None
            result = self._executor(config, _push_event, scan_control=control, interaction_fn=_web_interaction_fn)

            result_status = result.get("status", "complete")

            if result_status == "pause":
                state.status = "paused"
                state.paused_at_stage = result.get("paused_at_stage")
                state.paused_at = datetime.now(tz=timezone.utc).isoformat()
                state.completed_at = state.paused_at
                # The executor already pushed a "scan_paused" WebSocket event.

            elif result_status == "stop":
                state.status = "stopped"
                state.paused_at_stage = result.get("paused_at_stage")
                state.completed_at = datetime.now(tz=timezone.utc).isoformat()
                # The executor already pushed a "scan_stopped" WebSocket event.

            else:
                # Normal completion
                state.status = "complete"
                state.completed_at = datetime.now(tz=timezone.utc).isoformat()
                state.output_dir = result.get("output_dir")

                completion_details = {
                    "output_dir": state.output_dir,
                    "app_id": result.get("app_id", ""),
                }
                if result.get("low_confidence"):
                    completion_details["low_confidence"] = True

                _push_event(ProgressEvent(
                    event_type="stage_complete",
                    stage="complete",
                    percentage=100,
                    message="Pipeline completed successfully",
                    details=completion_details,
                ))

        except Exception as exc:
            logger.exception("Run %s failed", run_id)
            # Persist traceback to file for debugging
            import traceback as _tb
            _err_file = Path(f".threatforest/run_error_{run_id}.log")
            try:
                _err_file.write_text(_tb.format_exc(), encoding="utf-8")
                logger.info("Error details written to %s", _err_file)
            except OSError:
                pass

            # Save pause state so the run appears as a resumable paused scan.
            # Completed nodes are tracked by the executor and stored on the
            # ScanControl; if available, we persist them so resume can skip
            # already-finished stages.
            run_dir_str = control.run_dir if control else None
            if run_dir_str:
                try:
                    from server.executor import _save_pause_state
                    completed = getattr(control, "completed_nodes", []) or []
                    _save_pause_state(
                        Path(run_dir_str), completed, "error", config,
                    )
                    logger.info(
                        "Saved pause state for crashed run %s (%d completed nodes)",
                        run_id, len(completed),
                    )
                except Exception:
                    logger.warning(
                        "Could not save pause state for run %s", run_id, exc_info=True,
                    )

            state.status = "paused"
            state.completed_at = datetime.now(tz=timezone.utc).isoformat()
            state.error = str(exc)

            stage_name = getattr(exc, "stage", "unknown")
            _push_event(ProgressEvent(
                event_type="error",
                stage=stage_name,
                percentage=state_percentage(state),
                message=f"Pipeline failed: {exc}",
                details={
                    "error": str(exc),
                    "stage": stage_name,
                    "resumable": run_dir_str is not None,
                },
            ))


def state_percentage(state: RunState) -> float:
    """Derive a rough percentage from the current run state."""
    if state.status == "complete":
        return 100.0
    if state.status == "failed":
        return 0.0
    return 0.0

File: ts/packages/server/src/run-manager.ts
/**
 * Run manager — TS port of `src/server/run_manager.py`.
 *
 * Manages pipeline run lifecycle and streams progress events to WebSocket
 * clients. The Python version spawns OS threads (GIL-bound) and bridges via
 * asyncio queues; in Node the engine's `runGraph` is already async, so a run is
 * a background promise and progress events flow through a per-run async queue
 * (an EventEmitter-backed buffer that WS handlers drain).
 *
 * Pause/stop are cooperative: a ScanControl flag the executor checks at stage
 * boundaries. Resume rebuilds a RunConfig with `skip_nodes` for completed stages
 * (the engine graph + per-threat skip-if-exists handle the actual resumption).
 */
import { randomUUID } from 'node:crypto';
import {
  type RunConfig,
  type RunState,
  RunStateSchema,
} from '@threatforest/types';
import { ApplicationRegistry } from './registry.js';

/** Progress event pushed to WS clients. Mirrors run_manager.ProgressEvent.to_dict(). */
export interface ProgressEvent {
  type: string;
  stage: string;
  percentage: number;
  message: string;
  details: Record<string, unknown>;
  server_ts: number;
}

/** Cooperative pause/stop signal (port of server/scan_control.ScanControl). */
export class ScanControl {
  private _pause = false;
  private _stop = false;
  runDir: string | null = null;
  requestPause(): void {
    this._pause = true;
  }
  requestStop(): void {
    this._stop = true;
  }
  get shouldInterrupt(): boolean {
    return this._pause || this._stop;
  }
  get pauseRequested(): boolean {
    return this._pause;
  }
  get stopRequested(): boolean {
    return this._stop;
  }
}

export type OrchestratorExecutor = (
  config: RunConfig,
  onProgress: (e: ProgressEvent) => void,
  control: ScanControl,
  interactionFn: ((reason: Record<string, unknown>) => Promise<string | null>) | null,
) => Promise<{ status: string; output_dir?: string; app_id?: string; error?: string }>;

const TERMINAL = new Set(['complete', 'stopped', 'failed']);

/**
 * A simple async queue: producers push, a consumer awaits next().
 * Exported so the WS handler can name the type returned by
 * `ProgressBroadcaster.subscribe` (required for declaration emit under composite).
 *
 * Each subscriber gets its OWN AsyncQueue, so events fan out to every connection
 * instead of being stolen by whichever consumer calls next() first.
 */
export class AsyncQueue<T> {
  private buffer: T[] = [];
  private resolvers: ((v: T) => void)[] = [];
  push(item: T): void {
    const r = this.resolvers.shift();
    if (r) r(item);
    else this.buffer.push(item);
  }
  next(): Promise<T> {
    const item = this.buffer.shift();
    if (item !== undefined) return Promise.resolve(item);
    return new Promise((resolve) => this.resolvers.push(resolve));
  }
}

/** A subscription returned by {@link ProgressBroadcaster.subscribe}. */
export interface ProgressSubscription {
  /** This subscriber's private queue — independent cursor, no event stealing. */
  queue: AsyncQueue<ProgressEvent>;
  /** Detach this subscriber (call on WS close). */
  unsubscribe(): void;
}

/**
 * Per-run progress fan-out. A single producer (`publish`) broadcasts each event
 * to ALL subscribers' private queues, so multiple WS connections (or a browser
 * tab + a reconnect mid-flight) each receive the full stream independently.
 *
 * Replaces the old single shared AsyncQueue, where a second consumer would steal
 * events and any one disconnect could starve the others.
 */
export class ProgressBroadcaster {
  private readonly subscribers = new Set<AsyncQueue<ProgressEvent>>();

  publish(event: ProgressEvent): void {
    for (const q of this.subscribers) q.push(event);
  }

  subscribe(): ProgressSubscription {
    const queue = new AsyncQueue<ProgressEvent>();
    this.subscribers.add(queue);
    return {
      queue,
      unsubscribe: () => {
        this.subscribers.delete(queue);
      },
    };
  }

  get subscriberCount(): number {
    return this.subscribers.size;
  }
}

/**
 * A one-shot awaitable an agent blocks on while waiting for the user's answer
 * to an interviewer question. `submitInteractionResponse` resolves it.
 */
class PendingInteraction {
  private resolver: ((v: string | null) => void) | null = null;
  readonly promise: Promise<string | null>;
  constructor() {
    this.promise = new Promise((resolve) => {
      this.resolver = resolve;
    });
  }
  resolve(value: string | null): void {
    this.resolver?.(value);
    this.resolver = null;
  }
}

export class RunManager {
  readonly activeRuns = new Map<string, RunState>();
  private readonly broadcasters = new Map<string, ProgressBroadcaster>();
  private readonly controls = new Map<string, ScanControl>();
  private readonly history = new Map<string, ProgressEvent[]>();
  /** Per-run pending interviewer prompt, awaiting a `submitInteractionResponse`. */
  private readonly pending = new Map<string, PendingInteraction>();

  constructor(private readonly executor: OrchestratorExecutor | null = null) {}

  startRun(config: RunConfig): string {
    if (!this.executor) {
      throw new Error('No orchestrator executor configured. Provide one via new RunManager(executor).');
    }
    const runId = randomUUID().replace(/-/g, '');
    const now = new Date().toISOString();
    const state: RunState = RunStateSchema.parse({
      run_id: runId,
      status: 'pending',
      config,
      started_at: now,
    });
    this.activeRuns.set(runId, state);
    this.broadcasters.set(runId, new ProgressBroadcaster());
    this.history.set(runId, []);
    const control = new ScanControl();
    this.controls.set(runId, control);

    // Background execution (no thread needed — runGraph is async).
    void this.execute(runId, config, control);
    return runId;
  }

  private async execute(runId: string, config: RunConfig, control: ScanControl): Promise<void> {
    const state = this.activeRuns.get(runId)!;
    state.status = 'running';
    const onProgress = (e: ProgressEvent): void => {
      this.history.get(runId)?.push(e);
      this.broadcasters.get(runId)?.publish(e);
    };
    // Interviewer HITL: when the agent needs input it calls this fn, which emits
    // an `awaiting_input` event and blocks on a promise that the `/respond`
    // endpoint resolves via `submitInteractionResponse`.
    const interactionFn = async (reason: Record<string, unknown>): Promise<string | null> => {
      const p = new PendingInteraction();
      this.pending.set(runId, p);
      onProgress({
        type: 'awaiting_input',
        stage: 'interviewer',
        percentage: 0,
        message: 'Awaiting user input',
        details: reason,
        server_ts: Date.now(),
      });
      try {
        return await p.promise;
      } finally {
        this.pending.delete(runId);
      }
    };
    try {
      const result = await this.executor!(config, onProgress, control, interactionFn);
      if (result.status === 'pause') {
        state.status = 'paused';
        state.paused_at = new Date().toISOString();
        // The executor returns the run-dir ROOT here (where pause_state.json
        // lives); resumeRun reuses it. Without this, output_dir stayed null and
        // the progress-page Resume had no run dir to resume from.
        state.output_dir = result.output_dir ?? null;
      } else if (result.status === 'stop') {
        state.status = 'stopped';
        state.completed_at = new Date().toISOString();
        state.output_dir = result.output_dir ?? null;
      } else {
        state.status = result.status === 'complete' ? 'complete' : 'failed';
        state.completed_at = new Date().toISOString();
        state.output_dir = result.output_dir ?? null;
        if (result.error) state.error = result.error;
      }
    } catch (err) {
      state.status = 'failed';
      state.completed_at = new Date().toISOString();
      state.error = (err as Error).message;
    } finally {
      // Sentinel so WS consumers know the stream is done.
      onProgress({
        type: 'run_complete',
        stage: state.status,
        percentage: 100,
        message: `Run ${state.status}`,
        details: {},
        server_ts: Date.now(),
      });
    }
  }

  /**
   * Subscribe a WS connection to a run's live progress. Each call returns an
   * independent queue + unsubscribe, so concurrent connections (and reconnects)
   * each receive the full event stream rather than competing for it.
   * Throws if the run is unknown (caller falls back to history replay).
   */
  subscribeProgress(runId: string): ProgressSubscription {
    const b = this.broadcasters.get(runId);
    if (!b) throw new Error(`Unknown run_id: ${runId}`);
    return b.subscribe();
  }

  getHistory(runId: string): ProgressEvent[] {
    return this.history.get(runId) ?? [];
  }

  /** The ScanControl for a run (exposes runDir), or undefined if unknown. */
  getControl(runId: string): ScanControl | undefined {
    return this.controls.get(runId);
  }

  /**
   * Deliver a user's interviewer answer to the blocked agent thread.
   * Throws KeyError-equivalent if the run has no pending interaction.
   */
  submitInteractionResponse(runId: string, text: string | null): void {
    const p = this.pending.get(runId);
    if (!p) throw new Error(`No pending interaction for run_id: ${runId}`);
    p.resolve(text);
  }

  /**
   * Drop the per-run broadcaster ONLY once the run is terminal. History is
   * retained so late reconnects can still replay. Called by a WS connection on
   * close — but must NOT tear down a still-running stream, or other live
   * subscribers (or a reconnecting tab) would be starved. While running, a
   * disconnecting client just unsubscribes its own queue; the broadcaster lives
   * on until the run completes.
   */
  cleanupRun(runId: string): void {
    if (this.isTerminal(runId)) {
      this.broadcasters.delete(runId);
    }
  }

  pauseRun(runId: string): void {
    const state = this.activeRuns.get(runId);
    if (!state) throw new Error(`Unknown run_id: ${runId}`);
    if (state.status !== 'pending' && state.status !== 'running') {
      throw new Error(`Run ${runId} cannot be paused (status: ${state.status})`);
    }
    // A run blocked at a human-in-the-loop gate is suspended INSIDE a node and
    // yields no stream events, so the cooperative interrupt is never polled —
    // setting status='pausing' here would strand the run in 'pausing' until the
    // user answers the gate. Reject with a clear message; the user should
    // respond to (or skip) the open gate, then pause at the next boundary.
    if (this.pending.has(runId)) {
      throw new Error(
        `Run ${runId} is waiting for your input at a review step; respond to or skip it before pausing.`,
      );
    }
    state.status = 'pausing';
    this.controls.get(runId)?.requestPause();
  }

  stopRun(runId: string): void {
    const state = this.activeRuns.get(runId);
    if (!state) throw new Error(`Unknown run_id: ${runId}`);
    if (state.status === 'paused') {
      state.status = 'stopped';
      state.completed_at = new Date().toISOString();
      return;
    }
    if (!['pending', 'running', 'pausing'].includes(state.status)) {
      throw new Error(`Run ${runId} cannot be stopped (status: ${state.status})`);
    }
    // Same gate caveat as pauseRun: a run blocked awaiting input can't honor a
    // cooperative stop until the gate is resolved. Reject clearly.
    if (this.pending.has(runId)) {
      throw new Error(
        `Run ${runId} is waiting for your input at a review step; respond to or skip it before stopping.`,
      );
    }
    this.controls.get(runId)?.requestStop();
  }

  /**
   * Resume a paused/stopped run by starting a fresh run that reuses the run dir
   * and skips the nodes that already completed. For a paused/stopped run,
   * `state.output_dir` is the run-dir ROOT (the executor returns runDir, not
   * runDir/output, on interrupt), which is where `pause_state.json` lives. We
   * read it from disk (the durable source of truth that survives a process
   * restart) to recover `completed_nodes` → `skip_nodes`, so the resumed run
   * replaces those stages with no-op skip nodes and continues from the first
   * not-yet-completed node.
   */
  resumeRun(runId: string): string {
    const state = this.activeRuns.get(runId);
    if (!state) throw new Error(`Unknown run_id: ${runId}`);
    if (state.status !== 'paused' && state.status !== 'stopped') {
      throw new Error(`Run ${runId} is not resumable (status: ${state.status})`);
    }
    const runDir = state.output_dir ?? state.config.resume_run_dir;
    if (!runDir) {
      throw new Error(`Run ${runId} has no run directory to resume from`);
    }
    const pause = new ApplicationRegistry().readPauseState(runDir);
    const resumeConfig: RunConfig = {
      ...state.config,
      resume_run_dir: runDir,
      skip_nodes: pause?.completed_nodes ?? state.config.skip_nodes ?? [],
    };
    return this.startRun(resumeConfig);
  }

  isTerminal(runId: string): boolean {
    const s = this.activeRuns.get(runId);
    return !!s && TERMINAL.has(s.status);
  }
}
